fixed dones bug, added team spawn, loot item shuffle

kywch opened the pull request:
* fixed the env to return correct dones, added a test
* copied the baselines' team_helper.py to nmmo/lib, and use it for team spawning (i.e., putting the same team in the same tile)
* when looting items, randomize the item to be processed. Later, we may want to monitor how often the high-level items get discarded and try to minimize wasting high-level items


NOTE: There is another team helper in the task module, as some tasks can benefit from a team helper. We may want to consolidate it into the new team helper.

kywch committed:
fixed dones bug, added team spawn, loot item shuffle

kywch committed:
made pylint use multi cores

kywch committed:
map gen test use separate map dir

daveey commented on nmmo/core/env.py:
```diff
@@ -25,13 +26,16 @@ class Env(ParallelEnv):
-    config: Default = nmmo.config.Default(), seed=None):
+               config: Default = nmmo.config.Default(),
+               seed = None,
+               team_helper: TeamHelper = None):
```
I don't like this approach, we're making env/realm depend on teams, but the only reason to do that is for spawning.

Could you just create baselines/lib/team/team_loader.py: 

class TeamLoader(SequentialLoader):
  __init__(team_helper)
  
 Then we can just modify nmmo_config.py to use that loader. That would make environment/realm agnostic to teams

kywch replied:
@daveey looking entity_manager.py, spawn code where loader is being used.

hmm, it's more tangled than I thought. Let me have a look.

kywch committed:
removed team_helper from env/realm

kywch commented:
* Modified to NOT pass team_helper to the env and realm
* Refactored SequentialLoader and PlayerManager.spawn() to use TeamLoader (or any other custom Loader)'s `get_spawn_position(agent_id)` function for specifying the spawn position of each agent 
* Included a working example of `TeamLoader` in the tests/test_team_spawn.py
  * `config.PLAYER_LOADER = lambda config: TeamLoader(config, team_helper)`

daveey commented on nmmo/lib/team_helper.py:
```diff
@@ -0,0 +1,25 @@
+from typing import Dict, List
```
do we need this here?

kywch replied:
I guess so?

daveey commented on nmmo/entity/entity_manager.py:
```diff
@@ -141,7 +141,7 @@ class PlayerManager(EntityGroup):
     super().__init__(realm)
     self.loader  = self.realm.config.PLAYER_LOADER
-    self.agents = None
+    self.agents: spawn.SequentialLoader = None
```
do we want to call this self._agent_loader or something like that?

kywch replied:
changed `self.agents` to `self._agent_loader`, 
also changed `self.loader` to `self.loader_class` because the names were too similar

kywch committed:
renamed agent_loader variables
